controller/controller.go

```diff
@@ -0,0 +1,147 @@
+package controller
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"github.com/dgrijalva/jwt-go"
+	"iotdashboard/dbmanager"
+	"golang.org/x/crypto/bcrypt"
+	"time"
+)
+
+//TODO: these probably shouldn't be global vars
+var PSQL dbmanager.DBManager
+var jwtKey []byte
+
+// key is the JWT, value is the expiration epoch time
+var blocklist map[string]int64
+
+type Claims struct {
+	NotBefore int64 `json:"nbf"`
+	jwt.StandardClaims
+}
+
+////////Errors
+var ErrExpiredToken = errors.New("Token is expired")
+
+func init() {
+	var err error
+	jwtKey, err = GenerateRandomToken(256)
+	if err != nil {
+		panic(err)
+	}
+	PSQL, err = dbmanager.New("postgres", "myPassword", "iot_dashboard")
+	if err != nil {
+		panic(err)
+	}
+
+	blocklist = make(map[string]int64)
+}
+
+func Login(email, password string) (string, error) {
+	// use user struct instead of email and password?
+
+	// validate CSRF
+
+	// validate basic auth
+	err := CheckUserCredentials(email, password)
+	if err != nil {
+		return "", err
+	}
+
+	// create and return JWT
+	var token string
+	token, err = CreateJWT(60)
+	if err != nil {
+		return "", err
+	}
+	return token, nil
+
+}
+
+func Logout(token string) error {
+	// validate CSRF
+
+	// validate JWT
+	exp, err := ValidateJWT(token)
+	if err != nil {
+		return err
+	}
+	// blocklist JWT
+	blocklist[token] = exp
+	return nil
+
+}
+
+// returns an Error if the credentials are not valid
+func CheckUserCredentials(email, password string) error {
+	hash, err := PSQL.GetUserPassHash(email)
+	if err != nil {
+		return err
+	}
+	return bcrypt.CompareHashAndPassword(hash, []byte(password))
+}
+
+func CreateJWT(validPeriod int64) (string, error) {
```

Comment:
Feels like all utility methods for creating/validating tokens can be moved to a separate package/file.